### PR TITLE
checkov: init at 1.0.674

### PR DESCRIPTION
**Pull request by anhdle14**
###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [x] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by symphorien** [comment]
Is there a reason to not put the python dependencies in pkgs/development/python-modules and pkgs/top-level/python-packages.nix ?

**Comment by anhdle14** [comment]
The 4 I put here is the only version that worked with the build. I tried to not override, but the build will fail with the versions in python-packages.

**Comment by symphorien** [comment]
I don't know what the policy is about duplicating python packages in full vs overriding, cc @jonringer ?

deep_merge and bc-python-hcl2 are not in nixpkgs, so they should go with other python modules.

In any case, you should try to run tests. If you leave `doCheck = false`, please also leave a comment explaining why it's not possible to run the tests.

**Comment by jonringer** [comment]
The more generic python packages should be added to `python-modules.nix`.

If specific version is necessary, then I would prefer to do an override specific to the application as so:
```
let
  py = python.override {
    packageOverrides = self: super: {
      aws-sam-translator = super.aws-sam-translator.overridePythonAttrs (oldAttrs: rec {
        version = "1.25.0";
        src = oldAttrs.src.override {
          inherit version;
          sha256 = "08756yl5lpqgrpr80f2b6bdcgygr37l6q1yygklcg9hz4yfpccav";
        };
      });

      flask = super.flask.overridePythonAttrs (oldAttrs: rec {
        version = "1.0.2";
        src = oldAttrs.src.override {
          inherit version;
          sha256 = "0j6f4a9rpfh25k1gp7azqhnni4mb4fgy50jammgjgddw1l3w0w92";
        };
      });

      cookiecutter = super.cookiecutter.overridePythonAttrs (oldAttrs: rec {
        version = "1.6.0";
        src = oldAttrs.src.override {
          inherit version;
          sha256 = "0glsvaz8igi2wy1hsnhm9fkn6560vdvdixzvkq6dn20z3hpaa5hk";
        };
      });
    };
  };

in

with py.pkgs;

buildPythonApplication rec {
   ....
```

Overall stance on pinning python packages:

The preference to handling pinning is to relax version bounds in the "install_requires" field. (could be in setup.py, pyproject.toml, requirements or others). In most cases, packages are still compatible with small API changes which may warrant a major version bump. We use test suites to verify that the package still works correctly. 

If the package is still incompatible with the latest major version, then the most proper way to handle this is make an issue with the upstream package to adopt the latest major version. Or if upstream is not very responsive, you are free patch the source to make it compatible.

In very few circumstances, two versions of the same package are allowed to exist if the packages are extremely difficult to package. Some examples of this are tensorflow, which has huge ecosystems built around it and is hard to package. Another is django, which has 2 actively developed versions, and large ecosystems built around each.

One exception to this is applications, due to the way `buildPythonApplication` and `toPythonApplication` functions work, the related derivations will not bleed dependencies between packages. If the package doesn't need to be imported by other python modules, then this package would be a good candidate to convert into application. You can look at https://github.com/NixOS/nixpkgs/blob/master/pkgs/tools/admin/awscli/default.nix as an example of using an overlay within a python application. 

Info on `buildPythonApplication` can be found [here](https://github.com/NixOS/nixpkgs/blob/master/doc/languages-frameworks/python.section.md#buildpythonapplication-function).
Info on `toPythonApplication` can be found [here](https://github.com/NixOS/nixpkgs/blob/master/doc/languages-frameworks/python.section.md#topythonapplication-function).

**Comment by anhdle14** [comment]
@jonringer I updated based on guideline. Only `dpath` I kept within `checkov/default.nix` because `dpath 2.0` will crash. Should I write `dpath/1.5.nix` fo it?

File: pkgs/development/python-modules/bc-python-hcl2/default.nix [hunk 1]
```diff
@@ -0,0 +1,45 @@
+{ lib, buildPythonPackage, fetchPypi, nose }:
+
+let
+  lark-parser = buildPythonPackage rec {
+    pname = "lark-parser";
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/bc-python-hcl2/default.nix:5]: Why did you not create a separate package for this?

File: pkgs/development/python-modules/bc-python-hcl2/default.nix [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+  propagatedBuildInputs =  [
+    lark-parser
+    nose
+  ];
+
+  pythonImportsCheck = [ "hcl2" ];
+
+  meta = with lib; {
+    description = "A parser for HCL2 written in Python using Lark.";
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/bc-python-hcl2/default.nix:33]:
```suggestion
    description = "A parser for HCL2 written in Python using Lark";
```

File: pkgs/development/python-modules/deep_merge/default.nix [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+buildPythonPackage rec {
+  pname = "deep_merge";
+  version = "0.0.4";
+
+  src = fetchPypi {
+    inherit pname version;
+    sha256 = "tUQV+Qk0xC4zQRTihky01OczWzStOW41rYYQyWBlpH4=";
+  };
+
+  propagatedBuildInputs = [
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/deep_merge/default.nix:12]: Isn't this just required for the tests?

File: pkgs/development/python-modules/bc-python-hcl2/default.nix [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+  version = "0.3.11";
+
+  src = fetchPypi {
+    inherit pname version;
+    sha256 = "VZhI1oJ2EDZGyz3iI6/KYvJq4BGafzR+rcSgHqlUDrA=";
+  };
+
+  propagatedBuildInputs =  [
+    lark-parser
+    nose
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/bc-python-hcl2/default.nix:27]: I this required at runtime?

File: pkgs/development/python-modules/deep_merge/default.nix [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+    nose
+  ];
+
+  doCheck = false;
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/deep_merge/default.nix:16]: Please add a pythonImportsCheck.

File: pkgs/top-level/all-packages.nix [hunk 6]
```diff
@@ -28908,4 +28908,6 @@ in
   lc3tools = callPackage ../development/tools/lc3tools {};
 
   zktree = callPackage ../applications/misc/zktree {};
+
+  checkov = callPackage ../development/tools/analysis/checkov {};
```
Comment (SuperSandro2000) [inline pkgs/top-level/all-packages.nix:28912]: Please sort this into a category.

File: pkgs/development/tools/analysis/checkov/default.nix [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+  pname = "checkov";
+  version = "1.0.674";
+  src = fetchFromGitHub {
+    owner = "bridgecrewio";
+    repo = pname;
+    rev = version;
+    sha256 = "/S8ic5ZVxA2vd/rjRPX5gslbmnULL7BSx34vgWIsheQ=";
+  };
+
+  disabled = python3.pkgs.pythonOlder "3.7";
```
Comment (SuperSandro2000) [inline pkgs/development/tools/analysis/checkov/default.nix:13]:
```suggestion
  disabled = python3Packages.pythonOlder "3.7";
```

File: pkgs/development/tools/analysis/checkov/default.nix [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+  # Wrap the executable so that the python packages are available
+  # it's just a shebang script which calls `python -m checkov "$@"`
+  postFixup = ''
+    wrapProgram $out/bin/checkov \
+      --set PYTHONPATH $PYTHONPATH
+  '';
+
+  meta = with lib; {
+    homepage = "https://github.com/bridgecrewio/checkov";
+    description = "Checkov is a static code analysis tool for infrastructure-as-code.";
```
Comment (SuperSandro2000) [inline pkgs/development/tools/analysis/checkov/default.nix:73]:
```suggestion
    description = "Static code analysis tool for infrastructure-as-code";
```

**Comment by SuperSandro2000** [comment]
Result of `nixpkgs-review pr 107028` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>7 packages built:</summary>
  <ul>
    <li>checkov</li>
    <li>python37Packages.bc-python-hcl2</li>
    <li>python37Packages.deep_merge</li>
    <li>python38Packages.bc-python-hcl2</li>
    <li>python38Packages.deep_merge</li>
    <li>python39Packages.bc-python-hcl2</li>
    <li>python39Packages.deep_merge</li>
  </ul>
</details>

**Comment by SuperSandro2000** [comment]
Result of `nixpkgs-review pr 107028` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>7 packages built:</summary>
  <ul>
    <li>checkov</li>
    <li>python37Packages.bc-python-hcl2</li>
    <li>python37Packages.deep_merge</li>
    <li>python38Packages.bc-python-hcl2</li>
    <li>python38Packages.deep_merge</li>
    <li>python39Packages.bc-python-hcl2</li>
    <li>python39Packages.deep_merge</li>
  </ul>
</details>

File: pkgs/development/python-modules/bc-python-hcl2/default.nix (hunk 1 above)
Comment (anhdle14) [reply pkgs/development/python-modules/bc-python-hcl2/default.nix:5]: bc-python-hcl2 doesn't work with lark-parser from Nix Store. Should I add a `python-modules/lark-parser/0.7.8.nix` ?

File: pkgs/development/python-modules/deep_merge/default.nix (hunk 5 above)
Comment (anhdle14) [reply pkgs/development/python-modules/deep_merge/default.nix:16]: When added pythonImportsCheck it is always failing: 

```bash
deep_merge.tests.test_merge.test_default_behaviors ... ok
deep_merge.tests.test_merge.test_big_merge ... FAIL
deep_merge.tests.test_merge.test_no_references
Ensure that read-only parameters aren't being modified. ... FAIL

======================================================================
FAIL: deep_merge.tests.test_merge.test_big_merge
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/nix/store/g80x8dysb6hnhrcjrpav7dmsb8jnwnyr-python3.8-nose-1.3.7/lib/python3.8/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/private/tmp/nix-build-python3.8-deep_merge-0.0.4.drv-0/deep_merge-0.0.4/deep_merge/tests/test_merge.py", line 60, in test_big_merge
    assert merge(a, b, c) == expected
AssertionError

======================================================================
FAIL: deep_merge.tests.test_merge.test_no_references
Ensure that read-only parameters aren't being modified.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/nix/store/g80x8dysb6hnhrcjrpav7dmsb8jnwnyr-python3.8-nose-1.3.7/lib/python3.8/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/private/tmp/nix-build-python3.8-deep_merge-0.0.4.drv-0/deep_merge-0.0.4/deep_merge/tests/test_merge.py", line 73, in test_no_references
    assert output == {"key1": {"key1a": "2"}, "key2": [3, 4, 4, 5]}
AssertionError

----------------------------------------------------------------------
Ran 3 tests in 0.013s

FAILED (failures=2)
Test failed: <unittest.runner.TextTestResult run=3 errors=0 failures=2>
error: Test failed: <unittest.runner.TextTestResult run=3 errors=0 failures=2>
error: --- Error ------------------------------------------------------------------------------------------------------------------------ nix-build
builder for '/nix/store/41dyki3vl4qp53vd6kl92lg1yrn2w797-python3.8-deep_merge-0.0.4.drv' failed with exit code 1; last 10 log lines:
    File "/private/tmp/nix-build-python3.8-deep_merge-0.0.4.drv-0/deep_merge-0.0.4/deep_merge/tests/test_merge.py", line 73, in test_no_references
      assert output == {"key1": {"key1a": "2"}, "key2": [3, 4, 4, 5]}
  AssertionError

  ----------------------------------------------------------------------
  Ran 3 tests in 0.013s

  FAILED (failures=2)
  Test failed: <unittest.runner.TextTestResult run=3 errors=0 failures=2>
  error: Test failed: <unittest.runner.TextTestResult run=3 errors=0 failures=2>
```

File: pkgs/development/python-modules/bc-python-hcl2/default.nix (hunk 4 above)
Comment (anhdle14) [reply pkgs/development/python-modules/bc-python-hcl2/default.nix:27]: I will change this to `buildInputs`.

Sorry, I didn't carefully choose the dependencies' type.

File: pkgs/development/python-modules/bc-python-hcl2/default.nix (hunk 1 above)
Comment (SuperSandro2000) [reply pkgs/development/python-modules/bc-python-hcl2/default.nix:5]: This or if the package is simple enough just overwrite the version and hash here.

Comment (symphorien) [reply pkgs/development/python-modules/bc-python-hcl2/default.nix:5]: Anyway, add a comment explaining why the version is pinned.

File: pkgs/development/python-modules/bc-python-hcl2/default.nix (hunk 4 above)
Comment (symphorien) [reply pkgs/development/python-modules/bc-python-hcl2/default.nix:27]: please use `checkInputs` for test only dependencies.

File: pkgs/development/python-modules/deep_merge/default.nix (hunk 5 above)
Comment (symphorien) [reply pkgs/development/python-modules/deep_merge/default.nix:16]: Hum, that does not look like a spurious failure. Can you open an issue upstream to ensure that it's not nixpkgs-specific?

Comment (anhdle14) [reply pkgs/development/python-modules/deep_merge/default.nix:16]: That is a 3 years old repo... But I will try.

**Comment by anhdle14** [comment]
@symphorien  @SuperSandro2000  I will fix all the commits one the changes are all agreed. 

---

I want to ask on another part when moving `propagatedBuildInputs` to `buildInputs`.

This is the problem of moving `lark-parser 0.7.8` from `propagatedBuildInputs` to `buildInputs` in `bc-python-hcl2/default.nix`.

- Both `checkov` and `bc-python-hcl2` share the `lark-parser` requirements and they have to be the same version to build successfully.
- After testing, I think if switch to `buildInputs` I may need to patch it twice (`checkov/default.nix`, and `bc-python-hcl2/default.nix`). So I think it is best to include a lark-parser pinned version at `0.7.8` or just use `propagatedBuildInputs`.

I need your comments on the findings. Thank you!

File: pkgs/development/python-modules/bc-python-hcl2/default.nix [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+buildPythonPackage rec {
+  pname = "bc-python-hcl2";
+  version = "0.3.11";
+
+  src = fetchPypi {
+    inherit pname version;
+    sha256 = "VZhI1oJ2EDZGyz3iI6/KYvJq4BGafzR+rcSgHqlUDrA=";
+  };
+
+  # Nose is required during build process, so can not use `testInputs`.
```
Comment (symphorien) [inline pkgs/development/python-modules/bc-python-hcl2/default.nix:25]:
```suggestion
  # Nose is required during build process, so can not use `checkInputs`.
```

File: pkgs/development/python-modules/deep_merge/default.nix [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+buildPythonPackage rec {
+  pname = "deep_merge";
+  version = "0.0.4";
+
+  src = fetchPypi {
+    inherit pname version;
+    sha256 = "tUQV+Qk0xC4zQRTihky01OczWzStOW41rYYQyWBlpH4=";
+  };
+
+  testInputs = [
```
Comment (symphorien) [inline pkgs/development/python-modules/deep_merge/default.nix:12]:
```suggestion
  checkInputs = [
```

**Comment by symphorien** [comment]
leaving runtime python dependencies in propagatedBuildInputs is fine.

File: pkgs/development/tools/analysis/checkov/default.nix [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+      inherit pname version;
+      sha256 = "SWYVtOqEI20Y4NKGEi3nSGmmDg+H4sfsZ4f/KGxINhs=";
+    };
+
+    doCheck = false;
+  };
+
+  python = pkgs.python3.buildEnv.override {
+    extraLibs = [ dpath ];
+  };
```
Comment (symphorien) [inline pkgs/development/tools/analysis/checkov/default.nix:30]:
```suggestion
```

you should use python3Packages everywhere in the file.

Comment (anhdle14) [reply pkgs/development/tools/analysis/checkov/default.nix:30]: Sorry but how do I should write python3Packages in L28? I have checked the docs and it seems that `buildEnv` is not in `python3Packages`

Comment (symphorien) [reply pkgs/development/tools/analysis/checkov/default.nix:30]: I don't think you need this buildEnv at all. Add runtime python dependencies like dpath to propagatedBuildInputs.
Anyway, python is at `python3Packages.python` so buildEnv is at `python3Packages.python.buildEnv` I think.

**Pull request merged**